models/identity.go

```diff
@@ -26,6 +26,163 @@ func (Identity) TableName() string {
 	return tableName
 }
 
+// GetAccountLinkingDomain returns a string that describes the account linking
+// domain. An account linking domain describes a set of Identity entities that
+// _should_ generally fall under the same User entity. It's just a runtime
+// string, and is not typically persisted in the database. This value can vary
+// across time.
+func GetAccountLinkingDomain(provider string) string {
+	if strings.HasPrefix(provider, "sso:") {
+		// when the provider ID is a SSO provider, then the linking
+		// domain is the provider itself i.e. there can only be one
+		// user + identity per identity provider
+		return provider
+	}
+
+	// otherwise, the linking domain is the default linking domain that
+	// links all accounts
+	return "default"
+}
+
+type AccountLinkingDecision = int
+
+const (
+	AccountExists AccountLinkingDecision = iota
+	CreateAccount
+	LinkAccount
+	MultipleAccounts
+)
+
+type AccountLinkingResult struct {
+	Decision AccountLinkingDecision
+
+	User       *User
+	Identities []*Identity
+
+	LinkingDomain string
+}
+
+// DetermineAccountLinking uses the provided data and database state to compute a decision on whether:
+// - A new User should be created (CreateAccount)
+// - A new Identity should be created (LinkAccount) with a UserID pointing to an existing user account
+// - Nothing should be done (AccountExists)
+// - It's not possible to decide due to data inconsistency (MultipleAccounts) and the caller should decide
+//
+// Errors signal failure in processing only, like database access errors.
+func DetermineAccountLinking(tx *storage.Connection, provider, sub, email, phone string) (AccountLinkingResult, error) {
+	var similarIdentities []*Identity
+
+	if terr := tx.Q().Eager().Where("provider = ? and id = ?", provider, sub).All(&similarIdentities); terr != nil {
```

Comment:
this should always only be a single row because we have a PK constraint on (provider, id)

Reply:
Yup this was just to illustrate, didn't want to use existing methods to make the queries clearer.